### ISSUE TITLE
TSServer returns stale code actions

### ISSUE THREAD

**Issue by mjbvz** [Microsoft/TypeScript#14405]
_From @jrieken on February 27, 2017 10:2_

related to https://github.com/Microsoft/vscode/issues/21060

In the aforementioned issue TypeScript return a stale code action suggesting to add an import even though it has been added already. I assume this is because diagnostics computation has been scheduled but has not yet happened. The TS code action provider should synchronise with that actual source of code action.

_Copied from original issue: Microsoft/vscode#21482_

**Comment by mjbvz** [comment]
I confirmed that this seems to be a bug on the TypeScript side. Here's a trace of the using a code action to fix an import in the document. The order is:

1. Request code action for fix.
2. Apply fix.
3. Request new code actions since we have not received diagnostic info to clear the fixed error yet

```
[Trace - 8:06:00 PM] Sending request: getCodeFixes (85). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
    "startLine": 109,
    "endLine": 109,
    "startOffset": 27,
    "endOffset": 43,
    "errorCodes": [
        2304
    ]
}
[Trace - 8:06:00 PM] Response received: getCodeFixes (85). Request took 6 ms. Success: true 
Result: [
    {
        "description": "Add FileChangesEvent to existing import declaration from vs/platform/files/common/files.",
        "changes": [
            {
                "fileName": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
                "textChanges": [
                    {
                        "start": {
                            "line": 16,
                            "offset": 72
                        },
                        "end": {
                            "line": 16,
                            "offset": 72
                        },
                        "newText": ", FileChangesEvent"
                    }
                ]
            }
        ]
    }
]
[Trace - 8:06:03 PM] Sending request: change (86). Response expected: no. Current queue length: 0
Arguments: {
    "file": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
    "line": 16,
    "offset": 72,
    "endLine": 16,
    "endOffset": 72,
    "insertString": ", FileChangesEvent"
}
[Trace - 8:06:03 PM] Sending request: format (87). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
    "line": 16,
    "offset": 1,
    "endLine": 17,
    "endOffset": 1
}
[Trace - 8:06:03 PM] Response received: format (87). Request took 4 ms. Success: true 
Result: []
[Trace - 8:06:03 PM] Sending request: getCodeFixes (88). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
    "startLine": 109,
    "endLine": 109,
    "startOffset": 27,
    "endOffset": 43,
    "errorCodes": [
        2304
    ]
}
[Trace - 8:06:03 PM] Response received: getCodeFixes (88). Request took 133 ms. Success: true 
Result: [
    {
        "description": "Add FileChangesEvent to existing import declaration from vs/platform/files/common/files.",
        "changes": [
            {
                "fileName": "/Users/matb/projects/vscode/src/vs/workbench/services/textfile/common/textFileEditorModelManager.ts",
                "textChanges": [
                    {
                        "start": {
                            "line": 16,
                            "offset": 90
                        },
                        "end": {
                            "line": 16,
                            "offset": 90
                        },
                        "newText": ", FileChangesEvent"
                    }
                ]
            }
        ]
    }
]
```

Since the `format` call response is received, I expect that the TSServer has the fully updated version of the document. The returned code action also has the correct new offset value which supports this

**Comment by mhegazy** [comment]
@mjbvz is this still an issue?

**Comment by mjbvz** [comment]
I haven't seen any recent reports of this. Closing as fixed in 2.5
